### DISCUSSION TITLE
How to Efficiently Filter Combined Querysets from Multiple Models?

### DISCUSSION THREAD

**Discussion by c0d33py** [carltongibson/django-filter#1599] [Q&A]
**HELP WANTED**:I have a Django view that combines and sorts querysets from three different models (Event, Subject, and Article) to create a feed. I'm using the sorted function along with chain to merge the querysets and sort them based on the 'created_at' attribute. The feed is then returned as a list.

However, I'm facing challenges when it comes to filtering the feed based on user-provided search and author parameters. Since the feed is in list form, I can't directly use the filter method from the Django QuerySet.

The `created_by__username__icontains` is not working. Only the `ModelChoiceField` is working along with it but is want to use `CharField` field with it.

OR 

If anybody have another solution to make the feed. Please share it.

forms.py
```py
class FeedFilterForm(forms.Form):
    author = forms.CharField( # ModelChoiceField
        label='Author',
        required=False,
        # queryset=User.objects.all(),
        widget=forms.TextInput(attrs={'placeholder': 'Author'}),
    )
```

Views.py
```py
class FeedView(View):
    form_class = YourFormClass
    template_name = 'your_template.html'

    def get_queryset(self):
        """
        Get the combined and sorted queryset from events, subjects, and articles.
        """
        events = Event.objects.all()
        subjects = Subject.objects.all()
        articles = Article.objects.all()

        feed = sorted(
            chain(articles, subjects, events),
            key=attrgetter('created_at'),
            reverse=True,
        )

        queryset = self.filter_queryset(feed)
        return list(queryset)

    def filter_queryset(self, queryset):
        form = self.form_class(self.request.GET)

        if form.is_valid():
            data = form.cleaned_data

            author = data.get('author')
            if author:
                queryset = filter(
                    lambda x: author
                    == getattr(x, 'created_by__username__icontains', ''),
                    queryset,
                )

        return queryset

    def get(self, request, *args, **kwargs):
        queryset = self.get_queryset()

        context = {
            'object_list': queryset,
        }

        return render(request, self.template_name, context)
```

**Comment by carltongibson** [comment]
You want to use `QuerySet.union()` https://docs.djangoproject.com/en/4.2/ref/models/querysets/#union

**Reply by c0d33py** [reply]
I have three different base models. each model have different field and columns `union` oprator is getting `each UNION query must have the same number of columns
LINE 1: ...Y "core_article"."created_at" DESC) UNION (SELECT "core_subj...` while run the queryset `queryset = articles.union(tickers, events)`

**Reply by carltongibson** [reply]
Yes, so you need to select columns so that the query sets match up... 

I just checked, a search for "how to use union django" provides a number of good looking results.
